src/main/java/dev/jbang/source/GroovyScriptSource.java

```diff
@@ -0,0 +1,60 @@
+package dev.jbang.source;
+
+import static dev.jbang.net.GroovyManager.resolveInGroovyHome;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import org.jboss.jandex.ClassInfo;
+
+import dev.jbang.cli.BaseBuildCommand;
+import dev.jbang.net.GroovyManager;
+
+public class GroovyScriptSource extends ScriptSource {
+
+	protected GroovyScriptSource(ResourceRef script, Function<String, String> replaceProperties) {
+		super(script, replaceProperties);
+	}
+
+	@Override
+	public List<String> getCompileOptions() {
+		return Collections.emptyList();
+	}
+
+	@Override
+	public String getCompilerBinary(String requestedJavaVersion) {
+		return resolveInGroovyHome("groovyc", getGroovyVersion());
+	}
+
+	@Override
+	public Predicate<ClassInfo> getMainFinder() {
+		return pubClass -> pubClass.method("main", BaseBuildCommand.STRINGARRAYTYPE) != null
+				|| pubClass.method("main") != null;
+	}
+
+	@Override
+	public List<String> getAllDependencies() {
+		final List<String> allDependencies = super.getAllDependencies();
+		final String groovyVersion = getGroovyVersion();
+		if (groovyVersion.startsWith("4.")) {
```

Comment:
To future proof this condition I'd recommend grabbing the first number from the version, parse it as an int an check if it's >= 4

Reply:
groovyMajorVersion added.

```java
                 int groovyMajorVersion = 3;
		Matcher matcher = Pattern.compile("(\\d)\\..*").matcher(version);
		if (matcher.find()) {
			groovyMajorVersion = Integer.parseInt(matcher.group(1));
		}
```